cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go

```diff
@@ -122,6 +133,50 @@ type ClusterConfiguration struct {
 	ClusterName string `json:"clusterName,omitempty"`
 }
 
+// DNSAddOnType defines string identifying DNS add-on types
+type DNSAddOnType string
+
+const (
+	// CoreDNS add-on type
+	CoreDNS DNSAddOnType = "CoreDNS"
+
+	// KubeDNS add-on type
+	KubeDNS DNSAddOnType = "kube-dns"
+)
+
+// DNS defines the DNS addon that should be used in the cluster
+type DNS struct {
+	// Type defines string DNS add-on to be used
+	Type DNSAddOnType `json:"type"`
+
+	// ImageRepository, if set this fields allows to override the default image repository
+	// defined in ClusterConfiguration
+	ImageRepository string `json:"imageRepository,omitempty"`
+
+	// Image specifies which container image to use for running CoreDNS (this field does will be ignored if the selected
+	// add-on type is kube-dns).
+	// If empty, kubeadm will automatically select CoreDNS image using image repository and the supported CoreDNS version.
+	// In case this value is set, kubeadm does not change automatically the CoreDNS version during upgrades.
+	Image string `json:"image,omitempty"`
```

Comment:
it might be a good idea for us to outline that a tag is included here as part of Image.
e.g. Image = "image:tag".

we might get some reports in the lines of "hey, my custom coredns version upgrade is not working correctly", but that would be a side effect of the flexibility here, so it's a trade off.